lib/core/CProgramCounters.cc

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/CProgramCounters.h>
+
+#include <core/CLogger.h>
+#include <core/CRapidJsonLineWriter.h>
+#include <core/CStatePersistInserter.h>
+#include <core/CStateRestoreTraverser.h>
+#include <core/CStringUtils.h>
+
+#include <rapidjson/document.h>
+#include <rapidjson/ostreamwrapper.h>
+
+#include <ostream>
+#include <string>
+
+namespace ml {
+namespace core {
+
+namespace {
+
+using TGenericLineWriter = core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>;
+
+const std::string NAME_TYPE("name");
+const std::string DESCRIPTION_TYPE("description");
+const std::string COUNTER_TYPE("value");
+
+//! Persistence tags
+const std::string KEY_TAG("a");
+const std::string VALUE_TAG("b");
+
+//! Helper function to add a string/int pair to JSON writer
+void addStringInt(TGenericLineWriter& writer,
+                  const std::string& name,
+                  const std::string& description,
+                  uint64_t counter) {
+    writer.StartObject();
+
+    writer.String(NAME_TYPE);
+    writer.String(name);
+
+    writer.String(DESCRIPTION_TYPE);
+    writer.String(description);
+
+    writer.String(COUNTER_TYPE);
+    writer.Uint64(counter);
+
+    writer.EndObject();
+}
+}
+
+CProgramCounters& CProgramCounters::instance() {
+    return ms_Instance;
+}
+
+CProgramCounters::TCounter& CProgramCounters::counter(counter_t::ECounterTypes counterType) {
+    size_t counterPos = static_cast<size_t>(counterType);
+    if (counterPos >= counter_t::NUM_COUNTERS) {
+        // If the enum values in ECounterTypes have been maintained in a contiguous block then logically this
+        // can only happen when restoring state that contains one or more counters that are unknown to this version of the application.
+        // As this doesn't indicate a problem with the analytics in the running application we simply log a warning.
+        // A dummy counter is returned in which to store the unknown counter.
+        LOG_WARN(<< "Bad index " << counterPos);
+        return ms_Instance.m_DummyCounter;
+    }
+    return ms_Instance.m_Counters[counterPos];
+}
+
+CProgramCounters::TCounter& CProgramCounters::counter(size_t index) {
```

Comment:
nit: couldn't this just call overload with enum or vice verse?

Reply:
Will do, thanks Tom.